modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag

```diff
@@ -21,553 +21,805 @@
+				});
+			} catch (e) {
+				log.error("Cannot grant permissions to role '" + currentRole +
+					"' to perform action '" + carbon.registry.actions.GET + "'");
+				log.error(e);
+				// continue work flow
+			}
+		}
+
+		// check permissions to perform life cycle action
+		try {
+			if (!permissions.isLCActionsPermitted(user.username, currentArtifact.path,
+					userManager)) {
+				log.error("Insufficient permissions to perform '" + action + "' on " + appType +
+					" for user " + user.username);
+				response.status = 401;
+				response.contentType = "application/json";
+				content.status = "Insufficient Permissions";
+				content.message.push("You do not have permissions to perform action '" + action +
+					"' on version " + currentArtifact.attributes.overview_version + " of '" +
+					currentArtifact.attributes.overview_name + "' " + appType);
+				response.content = stringify(content);
+				return;
+			}
+		} catch (e) {
+			// this exception is thrown by
+			// org.wso2.carbon.user.core.AuthorizationManager.isRoleAuthorized(roleName,
+			// 										resourceId, action) throws UserStoreException
+			log.error("Cannot retrieve permissions for action '" + action + "' on artifact '" +
+				currentArtifact.path + "' for user " + user.username);
+			log.error(e);
+			response.status = 500;
+			response.contentType = "application/json";
+			content.status = "Error";
+			content.message = "Cannot retrieve permissions for action '" + action + "'";
+			response.content = stringify(content);
+			return;
+		}
+
+		// retrieve previously published web app versions (same provider & name)
+
+		// If everything went correctly in previous executions, previousArtifacts array should
+		// contain only 1 item. However due to some failure previousArtifacts can contain
+		// multiple items, that means there are more than 1 published version of this app.
+		// In order to handle these situation previousArtifacts array is processed as a
+		// multi-item array in the following code.
+		var previousArtifacts = [];
+
+		try {
+			previousArtifacts = artifactManager.find(function (artifact) {
+				var provider = (artifact.attributes.overview_provider ==
+				currentArtifact.attributes.overview_provider);
+				var name = (artifact.attributes.overview_name ==
+				currentArtifact.attributes.overview_name);
+				return (provider && name && (artifact.lifecycleState == 'Published'));
+			}, null);
+		} catch (e) {
+			log.error("Cannot retrieve previously published " + appType + " " +
+				currentArtifact.attributes.overview_provider + "-" +
+				currentArtifact.attributes.overview_provider);
+			log.error(e);
+			// continue work flow
+		}
+
+		var operationResult; // {isSuccess: true/false, message: ""}
+
+		// for all previously published apps when publishing or re-publishing current app
+		// unpublish previously published version of this app
+		if (((action == "Publish") || (action == "Re-Publish")) && (previousArtifacts.length > 0)) {
```

Comment:
Better to use "Publish" and "Re-Publish" as constant. There are number of usage of those in this class.